File: KV_DS_Assignment_Module20_Forecast_Airlines.py
# -*- coding: utf-8 -*-
"""
Created on Fri Jul 17 05:49:30 2020

@author: KavithaV
"""

import pandas as pd
airlines = pd.read_excel("D:/Data science/Module 20 - Forecast and Time series/CocaCola_Sales_Rawdata (1)/Airlines Data Set (1).xlsx")
import numpy as np
# Data Preparation
airlines["t"] = np.arange(1,97)

airlines["t_squared"] = airlines["t"]*airlines["t"]
airlines.columns
airlines["log_passengers"] = np.log(airlines["Passengers"])

##Extracting month column from timestamp
airlines['month'] = airlines['Month'].dt.strftime('%b')
# Dummy variables for Seasons
months =['Jan','Feb','Mar','Apr','May','Jun','Jul','Aug','Sep','Oct','Nov','Dec'] 

month_dummies = pd.DataFrame(pd.get_dummies(airlines['month']))
airlines = pd.concat([airlines,month_dummies],axis = 1)

# Partition the data
airlines.iloc[:, 1].plot() # Timeplot
Train = airlines.head(70)
Test = airlines.tail(26)

####################### L I N E A R ##########################
import statsmodels.formula.api as smf 

linear_model = smf.ols('Passengers ~ t', data=Train).fit()
pred_linear =  pd.Series(linear_model.predict(pd.DataFrame(Test['t'])))
rmse_linear = np.sqrt(np.mean((np.array(Test['Passengers'])-np.array(pred_linear))**2))
rmse_linear
#Out[44]: 48.30985585336209
##################### Exponential ##############################

Exp = smf.ols('log_passengers ~ t', data=Train).fit()
pred_Exp = pd.Series(Exp.predict(pd.DataFrame(Test['t'])))
rmse_Exp = np.sqrt(np.mean((np.array(Test['Passengers'])-np.array(np.exp(pred_Exp)))**2))
rmse_Exp
#Out[48]: 43.47847070446902
#################### Quadratic ###############################

Quad = smf.ols('Passengers ~ t + t_squared',data=Train).fit()
pred_Quad = pd.Series(Quad.predict(Test[["t","t_squared"]]))
rmse_Quad = np.sqrt(np.mean((np.array(Test['Passengers'])-np.array(pred_Quad))**2))
rmse_Quad
#Out[53]: 43.89814496742656
################### Additive seasonality ########################

add_sea = smf.ols('Passengers ~ Jan+Feb+Mar+Apr+May+Jun+Jul+Aug+Sep+Oct+Nov', data=Train).fit()
pred_add_sea = pd.Series(add_sea.predict(Test[['Jan','Feb','Mar','Apr','May','Jun','Jul','Aug','Sep','Oct','Nov']]))
rmse_add_sea = np.sqrt(np.mean((np.array(Test['Passengers'])-np.array(pred_add_sea))**2))
rmse_add_sea
#Out[90]: 124.97569951828822
################## Multiplicative Seasonality ##################

Mul_sea = smf.ols('log_passengers ~ Jan+Feb+Mar+Apr+May+Jun+Jul+Aug+Sep+Oct+Nov',data = Train).fit()
pred_Mult_sea = pd.Series(Mul_sea.predict(Test))
rmse_Mult_sea = np.sqrt(np.mean((np.array(Test['Passengers'])-np.array(np.exp(pred_Mult_sea)))**2))
rmse_Mult_sea
#Out[96]: 129.6291447652631
################## Additive Seasonality Quadratic Trend ############################

add_sea_Quad = smf.ols('Passengers ~ t+t_squared+Jan+Feb+Mar+Apr+May+Jun+Jul+Aug+Sep+Oct+Nov',data=Train).fit()
pred_add_sea_quad = pd.Series(add_sea_Quad.predict(Test[['Jan','Feb','Mar','Apr','May','Jun','Jul','Aug','Sep','Oct','Nov','t','t_squared']]))
rmse_add_sea_quad = np.sqrt(np.mean((np.array(Test['Passengers'])-np.array(pred_add_sea_quad))**2))
rmse_add_sea_quad 
#Out[102]: 30.39304289591171
################## Multiplicative Seasonality and Exponential ###########

Mul_Add_sea = smf.ols('log_passengers ~ t+Jan+Feb+Mar+Apr+May+Jun+Jul+Aug+Sep+Oct+Nov',data = Train).fit()
pred_Mult_add_sea = pd.Series(Mul_Add_sea.predict(Test))
rmse_Mult_add_sea = np.sqrt(np.mean((np.array(Test['Passengers'])-np.array(np.exp(pred_Mult_add_sea)))**2))
rmse_Mult_add_sea 
#Out[106]: 11.724791415163637
################## Testing #######################################

data = {"MODEL":pd.Series(["rmse_linear","rmse_Exp","rmse_Quad","rmse_add_sea","rmse_add_sea_quad","rmse_Mult_sea","rmse_Mult_add_sea"]),"RMSE_Values":pd.Series([rmse_linear,rmse_Exp,rmse_Quad,rmse_add_sea,rmse_add_sea_quad,rmse_Mult_sea,rmse_Mult_add_sea])}
table_rmse=pd.DataFrame(data)
table_rmse

# so rmse_Mult_add_sea has the least value among the models prepared so far 

File: KV_DS_Assignment_Module20_Forecast_Airlines.R
# Load the Dataset
library(readr)
library(readxl)
airlines <- read_xlsx("D:/Data science/Module 20 - Forecast and Time series/CocaCola_Sales_Rawdata (1)/Airlines Data Set (1).xlsx") # read the airlines data
View(airlines) 

windows()
plot(airlines$Passengers, type="o")

# Data Preparation
airlines["t"] <- 1:96
View(airlines)

airlines["t_square"] <- airlines["t"] * airlines["t"]
airlines["log_Passengers"] <- log(airlines["Passengers"])
attach(airlines)

# So creating 12 dummy variables 
X <- data.frame(outer(rep(month.abb, length = 96), month.abb,"==") + 0 )# Creating dummies for 12 months
View(X)

colnames(X) <- month.abb # Assigning month names 
View(X)

airlines <- cbind(airlines, X)
View(airlines)

colnames(airlines)

# Partition the time series
train <- airlines[1:70, ]
test <- airlines[71:96, ]


########################### LINEAR MODEL #############################
linear_model <- lm(Passengers ~ t, data=train)
summary(linear_model)

linear_pred <- data.frame(predict(linear_model,interval='predict', newdata =test))
View(linear_pred)

rmse_linear <- sqrt(mean((test$Passengers - linear_pred$fit)^2, na.rm = T))
rmse_linear # 48.30986


######################### Exponential #################################

expo_model <- lm(log_Passengers ~ t, data=train)
summary(expo_model)

expo_pred <- data.frame(predict(expo_model, interval='predict', newdata=test))
View(expo_pred)

rmse_expo<-sqrt(mean((test$Passengers - exp(expo_pred$fit))^2, na.rm = T))
rmse_expo # 43.47847

######################### Quadratic ####################################

Quad_model <- lm(Passengers ~ t + t_square, data=train)
summary(Quad_model)
Quad_pred <- data.frame(predict(Quad_model, interval='predict', newdata=test))

rmse_Quad <- sqrt(mean((test$Passengers - Quad_pred$fit)^2, na.rm=T))
rmse_Quad # 43.89814

######################### Additive Seasonality #########################

sea_add_model <- lm(Passengers ~ Jan+Feb+Mar+Apr+May+Jun+Jul+Aug+Sep+Oct+Nov, data=train)
summary(sea_add_model)

sea_add_pred <- data.frame(predict(sea_add_model, newdata=test, interval='predict'))

rmse_sea_add <- sqrt(mean((test$Passengers - sea_add_pred$fit)^2, na.rm = T))
rmse_sea_add # 124.9757

######################## Multiplicative Seasonality #########################

multi_sea_model <- lm(log_Passengers ~ Jan+Feb+Mar+Apr+May+Jun+Jul+Aug+Sep+Oct+Nov, data = train)
summary(multi_sea_model)

multi_sea_pred <- data.frame(predict(multi_sea_model, newdata=test, interval='predict'))

rmse_multi_sea <- sqrt(mean((test$Passengers - exp(multi_sea_pred$fit))^2, na.rm = T))
rmse_multi_sea # 129.6291

######################## Additive Seasonality with Linear trend #################

Add_sea_Linear_model <- lm(Passengers ~ t+Jan+Feb+Mar+Apr+May+Jun+Jul+Aug+Sep+Oct+Nov,data=train)
summary(Add_sea_Linear_model)
Add_sea_Linear_pred<-data.frame(predict(Add_sea_Linear_model,interval='predict',newdata=test))
rmse_Add_sea_Linear<-sqrt(mean((test$Passengers-Add_sea_Linear_pred$fit)^2,na.rm=T))
rmse_Add_sea_Linear # 34.50209

######################## Additive Seasonality with Quadratic Trend #################

Add_sea_Quad_model<-lm(Passengers ~ t+t_square+Jan+Feb+Mar+Apr+May+Jun+Jul+Aug+Sep+Oct+Nov,data=train)
summary(Add_sea_Quad_model)

Add_sea_Quad_pred<-data.frame(predict(Add_sea_Quad_model,interval='predict',newdata=test))
rmse_Add_sea_Quad<-sqrt(mean((test$Passengers-Add_sea_Quad_pred$fit)^2,na.rm=T))
rmse_Add_sea_Quad # 30.39304

######################## Multiplicative Seasonality Linear trend ##########################

multi_add_sea_model<-lm(log_Passengers ~ t+Jan+Feb+Mar+Apr+May+Jun+Jul+Aug+Sep+Oct+Nov,data = train)
summary(multi_add_sea_model) 

multi_add_sea_pred<-data.frame(predict(multi_add_sea_model,newdata=test,interval='predict'))
rmse_multi_add_sea<-sqrt(mean((test$Passengers-exp(multi_add_sea_pred$fit))^2,na.rm = T))
rmse_multi_add_sea # 11.72479

###### Preparing table on model and it's RMSE values 

table_rmse <- data.frame(c("rmse_linear","rmse_expo","rmse_Quad","rmse_sea_add","rmse_multi_sea","rmse_Add_sea_Quad","rmse_multi_add_sea"),c(rmse_linear,rmse_expo,rmse_Quad,rmse_sea_add,rmse_multi_sea,rmse_Add_sea_Quad,rmse_multi_add_sea))
View(table_rmse)

colnames(table_rmse) <-c ("model","RMSE")
View(table_rmse)

############### Multiplicative Seasonality Linear trend has least RMSE value############################# Additive seasonality with Quadratic Trend has least RMSE value

new_model <-lm(log_Passengers ~ t+Jan+Feb+Mar+Apr+May+Jun+Jul+Aug+Sep+Oct+Nov,data = train)
summary(new_model)

resid <- residuals(new_model)
resid[1:70]

# Predict
library(readr)
new_data <- read_csv(file.choose())
View(new_data)

pred_new <- data.frame(predict(new_model, newdata=new_data, interval = 'predict'))
View(pred_new)
pred_new$fit
